File: dotnet/src/Pactman.NonprofitCheckPlus/Models/OrganizationType.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Pactman.NonprofitCheckPlus.Models
{
    /// <summary>
    /// One Publication 78 deductibility entry from a nonprofit's <c>organization_types</c>.
    /// </summary>
    /// <remarks>
    /// Fields a newer API version adds stay readable through
    /// <see cref="DataObject.Get(string)"/>, exactly as on the organization itself.
    /// </remarks>
    public sealed class OrganizationType : DataObject
    {
        /// <summary>Initializes the entry from a decoded JSON object.</summary>
        /// <param name="element">The entry as the API sent it.</param>
        public OrganizationType(JsonElement element)
            : base(element)
        {
        }

        /// <summary>The Publication 78 organization type code.</summary>
        public string? Type => GetString("organization_type");

        /// <summary>The deductibility limitation the IRS records for this type.</summary>
        public string? DeductibilityLimitation => GetString("deductibility_limitation");

        /// <summary>The deductibility status, in the IRS's own words.</summary>
        public string? DeductibilityStatusDescription => GetString("deductibility_status_description");

        /// <summary>
        /// Reads a deductibility list, dropping entries the API could not resolve.
        /// </summary>
        /// <remarks>
        /// The API can send a null in the list, for a Publication 78 row it cannot
        /// resolve. Those are dropped rather than handed on, so every entry returned is a
        /// real object and the positions are renumbered from zero.
        /// </remarks>
        internal static IReadOnlyList<OrganizationType> ListFrom(JsonElement? element)
        {
            if (!element.HasValue || element.Value.ValueKind != JsonValueKind.Array)
            {
                return Array.Empty<OrganizationType>();
            }

            var entries = new List<OrganizationType>();

            foreach (var entry in element.Value.EnumerateArray())
            {
                if (entry.ValueKind == JsonValueKind.Object)
                {
                    entries.Add(new OrganizationType(entry));
                }
            }

            return entries;
        }
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/models/OrganizationType.java
package org.pactman.nonprofitcheckplus.models;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/** A deductibility entry from IRS Publication 78. */
public final class OrganizationType extends DataObject {

    /**
     * Initializes the entry from a decoded JSON object.
     *
     * @param fields the entry as the API sent it.
     */
    public OrganizationType(Map<String, Object> fields) {
        super(fields);
    }

    /**
     * Reads a decoded {@code organization_types} value as a list of entries.
     *
     * <p>Returns an empty list when the value is absent, JSON null, or not an
     * array, so it is always safe to iterate.
     *
     * <p>The API can also send a null inside the list, for a Publication 78 row
     * it cannot resolve. Those are dropped rather than handed on, so every entry
     * this returns is a real object and the positions are renumbered from zero —
     * read {@code get("organization_types")} when the gaps themselves matter.
     *
     * @param value the decoded field value.
     * @return the entries the API sent.
     */
    public static List<OrganizationType> listFrom(Object value) {
        if (!(value instanceof List)) {
            return Collections.emptyList();
        }

        List<?> raw = (List<?>) value;
        List<OrganizationType> entries = new ArrayList<>(raw.size());

        for (Object entry : raw) {
            if (entry instanceof Map) {
                entries.add(new OrganizationType(asFields(entry)));
            }
        }

        return Collections.unmodifiableList(entries);
    }

    /**
     * The deductibility text for this entry.
     *
     * @return the text, or {@code null} when the API returned none.
     */
    public String getOrganizationType() {
        return getString("organization_type");
    }

    /**
     * The deductibility limitation, such as {@code 50%}.
     *
     * @return the limitation, or {@code null} when the API returned none.
     */
    public String getDeductibilityLimitation() {
        return getString("deductibility_limitation");
    }

    /**
     * The deductibility status code, such as {@code PC}.
     *
     * @return the status description, or {@code null} when the API returned none.
     */
    public String getDeductibilityStatusDescription() {
        return getString("deductibility_status_description");
    }
}
